### ISSUE TITLE
Ts does not enforce type provision for Generic React Component

### ISSUE THREAD

**Issue by minseoksuh** [microsoft/TypeScript#46355]
# Bug Report

Hi

I researched the way to create a Generic Component.

![typeError1](https://user-images.githubusercontent.com/54365519/114980812-00437200-9ec8-11eb-9374-957c4fae313a.png)

And it does what I want it to do, forcing types for values inside the component.

However the issue I found buggy was that,

<img width="987" alt="Screen Shot 2021-04-16 at 3 30 27 PM" src="https://user-images.githubusercontent.com/54365519/114981281-b27b3980-9ec8-11eb-9e6a-ec5bee4fcef2.png">

unlike normal Generic function which catches

> Generic type 'FuncProps<T>' requires 1 type argument(s).ts(2314)

generic React Component does not catch when Type is not provided.

TS just infers the types of the options, and does not force Type input.

<img width="906" alt="Screen Shot 2021-04-19 at 11 39 10 AM" src="https://user-images.githubusercontent.com/54365519/115174524-5dc40280-a104-11eb-8bc0-22e3950ffb8a.png">

### 🔎 Search Terms

- generic type
- generic react component

### 🕗 Version & Regression Information

- This is a crash
- This changed between versions ______ and _______
- This is the behavior in every version I tried, and I reviewed the FAQ for entries about _________
- I was unable to test this on prior versions because _______

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/PTAEFkE9QYwewCYFNQHM5IM6gBZIE5JA)

### 💻 Code

```ts
// We can quickly address your report if:
//  - The code sample is short. Nearly all TypeScript bugs can be demonstrated in 20-30 lines of code!
//  - It doesn't use external libraries. These are often issues with the type definitions rather than TypeScript bugs.
//  - The incorrectness of the behavior is readily apparent from reading the sample.
// Reports are slower to investigate if:
//  - We have to pare too much extraneous code.
//  - We have to clone a large repo and validate that the problem isn't elsewhere.
//  - The sample is confusing or doesn't clearly demonstrate what's wrong.
```

### 🙁 Actual behavior

Generic Type React component does not force type provision, 
and infers type from value provided.

### 🙂 Expected behavior

Generic component should cause type error when type is not provided.

**Comment by minseoksuh** [comment]
@andrewbranch Hey thanks for the explanation.

It may be just that my understanding of this matter is just wrong,
But I dont think we are talking about the same issue here.
Please bear with me one more time.

there are times you want ts to infer from your inputs
but there are also times you want to tell ts what kind of data you are dealing with so you can perform specific actions on it.

<img width="987" alt="Screen Shot 2021-04-16 at 3 30 27 PM" src="https://user-images.githubusercontent.com/54365519/114981281-b27b3980-9ec8-11eb-9e6a-ec5bee4fcef2.png">

In the above code,
ts throws type error
`Generic type 'FuncProps<T>' requires 1 type argument(s).`
because it has to know about what T will be in order to understand what kind of operations is allowed for props.value.

```
interface FuncProps<T> {
    value: T;
}

const stringFunc = (props: FuncProps<string>) => {
    props.value.substring(0);
}
```

Because I provided what my generic type is above, I can safely call operations specific to that type.

```
stringFunc({value: 2222})
```
and above will cause type error because the type of value(number) does not work with how I defined stringFunc.

And I am saying what is the difference between React generic Function component and normal generic functions?

If I am trying to make a dropdown for which I want to ensure the value I am dealing with will be a string, so I can use string specific functions inside of it, the value type should not be inferred from the input it is provided with.
The dropdown should not accept wrong input value in the first place.

**Comment by minseoksuh** [comment]
@andrewbranch 

I finally understood where my logical hole was.

```
interface Props<T>{
    value: T;
}

const testFunc = <T>(props: Props<T>){

    console.log({props})
}

testFunc({value: "1111"})
```
the above code works exactly like generic react components. It infers value from input.

As soon as I make my React Component Generic in the case of dropdown I am already saying type of value is a variable.

Thank you for the explanation.
